chain/chain/src/store.rs

```diff
@@ -2066,19 +2031,168 @@ impl<'a> ChainStoreUpdate<'a> {
                 self.clear_chunk_data(min_chunk_height)?;
             }
             GCMode::StateSync => {
-                // 7. State Sync clearing
-                // 7a. Delete blocks with current height (ColBlockPerHeight)
-                store_update.delete(ColBlockPerHeight, &index_to_bytes(height));
-                self.chain_store.block_hash_per_height.cache_remove(&index_to_bytes(height));
-                // 7b. Delete from ColBlockHeight - don't do because: block sync needs it + genesis should be accessible
-
+                // 8. State Sync clearing
+                // Don't delete from ColBlockHeight because: block sync needs it + genesis should be accessible
                 // Chunks deleted separately
             }
         };
         self.merge(store_update);
         Ok(())
     }
 
+    fn get_gc_count(&self, col: DBCol) -> u64 {
+        if let Some(count) = self.gc_count.get(&(col as u64)) {
+            *count
+        } else if let Ok(Some(count)) = self
+            .store()
+            .get(DBCol::ColGCCount, &borsh::ser::BorshSerialize::try_to_vec(&(col as u64)).unwrap())
+        {
+            borsh::de::BorshDeserialize::try_from_slice(&count).unwrap()
+        } else {
+            0
+        }
+    }
+
+    fn inc_gc(&mut self, col: DBCol) {
+        let new_count: u64 = self.get_gc_count(col) + 1;
+        self.gc_count.insert(col as u64, new_count);
+    }
+
+    pub fn gc_col_block_per_height(
+        &mut self,
+        block_hash: &CryptoHash,
+        height: BlockHeight,
+        epoch_id: &EpochId,
+    ) -> Result<(), Error> {
+        let mut store_update = self.store().store_update();
+        let epoch_to_hashes_ref = self.get_all_block_hashes_by_height(height)?;
+        let mut epoch_to_hashes = epoch_to_hashes_ref.clone();
+        let hashes = epoch_to_hashes.get_mut(epoch_id).expect("current epoch id should exist");
+        hashes.remove(&block_hash);
+        if hashes.is_empty() {
+            epoch_to_hashes.remove(epoch_id);
+        }
+        if epoch_to_hashes.is_empty() {
+            store_update.delete(ColBlockPerHeight, &index_to_bytes(height));
+            self.chain_store.block_hash_per_height.cache_remove(&index_to_bytes(height));
+        } else {
+            store_update.set_ser(ColBlockPerHeight, &index_to_bytes(height), &epoch_to_hashes)?;
+            self.chain_store
+                .block_hash_per_height
+                .cache_set(index_to_bytes(height), epoch_to_hashes);
+        }
+        self.inc_gc(DBCol::ColBlockPerHeight);
+        self.merge(store_update);
+        Ok(())
+    }
+
+    pub fn gc_col_state(&mut self, tries: ShardTries, keys: Vec<Vec<u8>>) {
+        let mut store_update = StoreUpdate::new_with_tries(tries);
+
+        for key in keys.iter() {
+            store_update.delete(DBCol::ColState, key.as_ref());
+            self.inc_gc(DBCol::ColState);
+        }
+
+        self.merge(store_update);
+    }
```

Comment:
`ColState` is cleared in several GCModes: Canonical, Fork and Sync. In Canonical and Fork ones, it implicitly happens in `revert_insertions` and `apply_deletions`. In Sync mode, it happens explicitly by removing all the records from `ColState`.

This method is confusing because it sounds like it covers all three modes. However, it's not.

What we can do to make things clearer:
1. Extend `gc_col_state` to clear ColState in Canonical and Fork modes as well. It might be annoying and looks like some over-engineering.
2. Get rid of `gc_col_state` and put this code into `reset_data_pre_state_sync` back.
3. Rename this method into `state_sync_gc_col_state` (or like that) and change panic message in `gc_col()`.

Any of those solutions is acceptable to me.